### PR TITLE
feat(tags): handle case-insensitive tags and remove orphans

### PR DESCRIPTION
**Pull request by laurent-laporte-pro**
What's new

- Change the search and update methods in `StudyMetadataRepository` class to handle case-insensitive tags and remove orphans.
- Update the Alembic migration script to handle case-insensitive tags.
- Change the filtering in the font-end to handle case-insensitive tags.

### CODE REVIEW

**Review by skamril — DISMISSED** [review]

File: alembic/versions/dae93f1d9110_populate_tag_and_study_tag_tables_with_.py [hunk 1]
```diff
@@ -23,6 +24,22 @@
 depends_on = None
 
 
+def _avoid_duplicates(tags: t.Iterable[str]) -> t.Sequence[str]:
+    """Avoid duplicate tags (case insensitive)"""
+
+    upper_tags = {tag.upper(): tag for tag in tags}
+    return list(upper_tags.values())
```
Comment (mabw-rte) [inline alembic/versions/dae93f1d9110_populate_tag_and_study_tag_tables_with_.py:31]: J'aurais preferer une optimisation du code: 
```python
def _avoid_duplicates(tags: t.Iterable[str]) -> t.Sequence[str]:
    """Avoid duplicate tags (case insensitive)"""

    return list({tag.upper() for tag in tags})
```

`set` gere automatiquement les tags dupliques

File: antarest/study/repository.py [hunk 2]
```diff
@@ -279,17 +280,23 @@ def delete(self, id_: str, *ids: str) -> None:
...
+        new_upper_tags = {tag.upper(): tag for tag in new_tags}
+        existing_tags = self.session.query(Tag).filter(func.upper(Tag.label).in_(new_upper_tags)).all()
+        for tag in existing_tags:
+            if tag.label.upper() in new_upper_tags:
+                new_upper_tags.pop(tag.label.upper())
+        study.tags = [Tag(label=tag) for tag in new_upper_tags.values()] + existing_tags
         self.session.merge(study)
         self.session.commit()
+        # Drop orphan tags
+        self.session.query(Tag).filter(~Tag.studies.any()).delete(synchronize_session=False)  # type: ignore
```
Comment (mabw-rte) [inline antarest/study/repository.py]: Add a comment for developers, so that they realize that if a `tag` is not associated with any study it will be deleted. So if `tags` are to be associated with object other than `Study`, this code must be updated.

**Review by mabw-rte — CHANGES_REQUESTED** [review]

File: alembic/versions/dae93f1d9110_populate_tag_and_study_tag_tables_with_.py (hunk 1 above)
Comment (mabw-rte) [reply alembic/versions/dae93f1d9110_populate_tag_and_study_tag_tables_with_.py:31]: si `tags` est vide la liste le sera egalement.

Comment (laurent-laporte-pro) [reply alembic/versions/dae93f1d9110_populate_tag_and_study_tag_tables_with_.py:31]: En fait, ce n'est pas la même chose. Dans ma version de `_avoid_duplicates`, je retourne la liste des tags inchangés (mais dédoublonnés en étant insensible à la casse), alors que dans ce que tu proposes, tu retournes la liste des tags en majuscules.

File: antarest/study/repository.py [hunk 3]
```diff
@@ -279,17 +280,24 @@ def delete(self, id_: str, *ids: str) -> None:
...
+        for tag in existing_tags:
+            if tag.label.upper() in new_upper_tags:
+                new_upper_tags.pop(tag.label.upper())
+        study.tags = [Tag(label=tag) for tag in new_upper_tags.values()] + existing_tags
         self.session.merge(study)
         self.session.commit()
+        # Delete any tag that is not associated with any study.
+        # Note: If tags are to be associated with objects other than Study, this code must be updated.
+        self.session.query(Tag).filter(~Tag.studies.any()).delete(synchronize_session=False)  # type: ignore
+        self.session.commit()
```
Comment (hdinia) [inline antarest/study/repository.py]: You do not account for whitespaces in this normalization logic, which can lead to unintended duplicates. For example:
` " Test","Test","Test "` =>  will all pass and return => `" TEST","Test","test "` 

Also inputs are mutated , and this is a bad practice regarding UX. if a user enters a tag as "NodeJs" expecting case sensitivity to be preserved, they might be surprised or confused to see it displayed as "NODEJS" in the application. 

The same issue arises with special chars, if a user inputs "&Test", you will end up with another duplicate, if we dont consider this a duplicate then its ok.

Please add whitespace trimming and special character normalization to the tag processing logic.

**Review by hdinia — CHANGES_REQUESTED** [review]

**Review by hdinia — CHANGES_REQUESTED** [review]
Here's a little feedback in the hope that it will help:

- User Input Mutation: Mutating user inputs can lead to unexpected UX issues, including perceived bugs and user confusion.

- Incomplete Normalization: The proposed normalization solution does not adequately address edge cases (whitespace)

Recommendations:

- Implement Validation: Instead of normalizing inputs, validate them for duplicates and format issues directly at the point of entry. We may need to resolve existing duplicates as well. In this case, normalization can help, or a database migration could be implemented to remove all duplicates. 
- This validation will occur both on the frontend (user input checks and instant feedback, as shown in the screenshot below) and on the backend (checking against database values).
> if needed I can commit the front-end part, let me know

Example:
image


- Preserve User Input Integrity: This approach maintains user input avoiding UX issues and bug-like behavior.

- Adopting a validation-centric strategy would significantly improve user experience by providing clear, immediate feedback (error messages) and avoiding the pitfalls of input mutation.

Could you provide further clarification on the rationale behind the current approach? I apologize if I've misunderstood any aspects of the implementation.

**Review by mabw-rte — DISMISSED** [review]

File: antarest/study/repository.py (hunk 3 above)
Comment (laurent-laporte-pro) [reply antarest/study/repository.py]: J'ai modifié le point d'accès de l'API `PUT /v1/studies/{uuid}` afin de normaliser les espaces des tags. Je préfère effectuer cette normalisation en amont, au niveau du endpoint, plutôt qu'en aval, au moment de l'enregistrement des tags dans la base de données. Cela permet notamment de vérifier la longueur des tags et de renvoyer une erreur 422 en cas de longueur incorrecte.

Je ne comprends pas ce que tu veux dire par "Also inputs are mutated". En fait, dans l'interface web, il serait nécessaire d'améliorer la déduplication des tags en les rendant insensibles à la casse et en normalisant les espaces. Actuellement, la déduplication fonctionne uniquement de manière sensible à la casse et sans normalisation des espaces. C'est quelque chose que je ne sais pas faire en React. Si tu peux t'en charger, tu es le bienvenu.

Par contre, je pense qu'il ne faut pas trop en faire au sujet des tags et que permettre les caractères spéciaux n'est pas pénalisant.

Comment (hdinia) [reply antarest/study/repository.py]:
```python
    def test_tag_capitalization_on_update(self, client: TestClient, user_access_token: str,
                                             study_id: str) -> None:
        """
        Test to handle specific edge cases like "test" mutating to "Test" or "node" to "Node".
        """

        # Test cases for specific mutations
        test_cases = [("test", "Test"), ("node", "Node")]
        for original_tag, expected_mutation in test_cases:
            res = client.put(
                f"/v1/studies/{study_id}",
                headers={"Authorization": f"Bearer {user_access_token}"},
                json={"tags": [original_tag]},
            )
            assert res.status_code == 200, res.json()
            actual = res.json()
            assert set(actual["tags"]) == {
                expected_mutation}, f"Expected {expected_mutation} but got {set(actual['tags'])}"
```

Tu comprendras mieux avec ce test qui reproduit l'anomalie. C'est un Edge case et cela ne break rien (juste bizarre niveau UX) mais au moins vous êtes au courant que c'est la.
> note: il y a aussi des cas ou l'API renvoie du UPPER alors que l'input etais en case "mixte"

Edit: on a debeug avec @laurent-laporte-pro, c'est un cas un peu aléatoire on ne reproduit pas en TU donc surement dû à l'environnement de dev, tester sur le serveur d'intégration c'est OK on ne reproduit pas.

**Review by hdinia — APPROVED** [review]
